ganttproject/src/net/sourceforge/ganttproject/task/algorithm/RecalculateTaskCompletionPercentageAlgorithm.java

```diff
@@ -43,29 +43,52 @@ public void run(Task task) {
     recalculateSupertaskCompletionPercentageBottomUp(task, facade);
   }
 
+  private class PropagatedDays {
+    public int myCompletedDays;
+    public long myPlannedDays;
+
+    PropagatedDays() {
+      myCompletedDays = 0;
+      myPlannedDays = 0;
+    };
+
+    PropagatedDays(int completedDays, long plannedDays) {
+      myCompletedDays = completedDays;
+      myPlannedDays = plannedDays;
+    };
+  }
+
   private void recalculateSupertaskCompletionPercentageBottomUp(Task task, TaskContainmentHierarchyFacade facade) {
-    while (task != null) {
-      recalculateSupertaskCompletionPercentage(task, facade);
-      task = facade.getContainer(task);
-    }
+     Task root = facade.getRootTask();
+     recalculateSupertaskCompletionPercentage(root, facade);
   }
 
-  private void recalculateSupertaskCompletionPercentage(Task task, TaskContainmentHierarchyFacade facade) {
+  private PropagatedDays recalculateSupertaskCompletionPercentage(Task task, TaskContainmentHierarchyFacade facade) {
+
     Task[] nestedTasks = facade.getNestedTasks(task);
+
+    int completedDays = 0;
+    long plannedDays = 0;
+
     if (nestedTasks.length > 0) {
-      int completedDays = 0;
-      long plannedDays = 0;
       for (int i = 0; i < nestedTasks.length; i++) {
         Task next = nestedTasks[i];
-        long nextDuration = next.isMilestone() ? 1 : next.getDuration().getLength();
-        completedDays += nextDuration * next.getCompletionPercentage();
-        plannedDays += nextDuration;
+        PropagatedDays propagatedDays = recalculateSupertaskCompletionPercentage(next, facade);
+        completedDays += propagatedDays.myCompletedDays;
+        plannedDays += propagatedDays.myPlannedDays;
       }
       int completionPercentage = plannedDays == 0 ? 0 : (int) (completedDays / plannedDays);
       task.setCompletionPercentage(completionPercentage);
+    } else {
+      long nextDuration = task.getDuration().getLength();
```

Comment:
I suggest returning `new PropagatedDays(nextDuration * task.getCompletionPercentage(), nextDuration)` right away and moving accumulating vars into the previous branch of if

Reply:
Any particular reason for this?
This way there have to be two return statements, one within each branch of the if-statement which leaves the method without an obvious return statement.

Reply:
Just to have less mutable vars. In fact, I would even remove else like below but that doesn't matter too much to me :)

```
if (nestedTasks.length == 0) {
  return new SubtreeCompletion....
}
long completedDays...
// other code from current if > 0 block